krabs/krabs/schema_locator.hpp

```diff
@@ -124,18 +124,34 @@ namespace krabs {
         void reset();
 
         /**
-         * Get the schema_locator singleton instance.
+         * Get the schema_locator instance for this thread.
          * WARNING: if compiled into managed code, do NOT get a reference
          * to this in a static initializer or it will cause loader lock.
          */
-        static schema_locator& get_instance()
+        static schema_locator& get_thread_instance()
         {
-            return singleton_;
+            if (index_ == 0)
+            {
+                scope_lock l(sync_);
+                if (index_ == 0) index_ = TlsAlloc();
+            }
+
+            if (index_ == TLS_OUT_OF_INDEXES)
+                throw std::runtime_error("TLS_OUT_OF_INDEXES");
+
+            auto tls_locator = static_cast<schema_locator*>(TlsGetValue(index_));
+            if (tls_locator == nullptr)
+            {
+                tls_locator = new schema_locator;
```

Comment:
Sorry I messed this the first time, but this is leaked. The Tls block is freed for you, but this is heap allocated and never freed. You’ll need to figure out the callback that is called before the thread dies so you can call delete in the pointer stored here.

Reply:
Maybe you can make and call a cleanup function in the per-thread that calls the function which pumps ETW. When the blocking ETW function returns, maybe you can call to reset or something then.

Reply:
I'll look into cleaning up the allocation when the blocking function returns. 
This seems more appropriate than a DLL_THREAD_DETACH approach - since we aren't allocating per-thread, just per tracing thread.

Reply:
Hmm. Actually this needs some more thought. schema_locator can be used outside of the processing thread - e.g. half of the tests. :-)

Reply:
I don't think that DLL_THREAD_DETACH is an option - since the C++ implementation is headers only.

Unless... we #define different implementations for C++ and C++/CLI wrapper...
We could even use the preferred/performant thread_local approach for C++ then.